src/smolagents/agents.py

```diff
@@ -1296,9 +1277,33 @@ def _step_stream(self, memory_step: ActionStep) -> Generator[ChatMessageStreamDe
         else:
             for tool_call in chat_message.tool_calls:
                 tool_call.function.arguments = parse_json_if_needed(tool_call.function.arguments)
-        yield from self.process_tool_calls(chat_message, memory_step)
+        tool_outputs = {}
+        got_final_answer = False
+        for output in self.process_tool_calls(chat_message, memory_step):
+            yield output
+            if isinstance(output, ToolOutput):
+                tool_outputs[output.id] = output
+                if output.is_final_answer:
+                    got_final_answer = True
+
+                    # Manage state variables
+                    if isinstance(output.output, str) and output.output in self.state.keys():
+                        output.output = self.state[output.output]
+
+        if len(tool_outputs) == 1:
+            final_output = list(tool_outputs.values())[0].output
+        else:
+            # ToolCallingAgent can return several final answers in parallel
+            final_output = [tool_outputs[k].output for k in sorted(tool_outputs.keys())]
```

Comment:
Just a small observation here: by sorting `tool_outputs.keys()`, we're effectively ordering the final outputs by `tool_call.id` (which is what `output.id` mirrors). This might not always reflect the original order in which the tool calls were made. If the logical or intended sequence depends on the call order rather than the ID's lexicographic order, would it make sense to preserve that instead?